man/man1/raxpol_ray_hdrs.1
.\" 
.\" Copyright (c) 2015, Gordon D. Carrie. All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\"
.Pp
.Dd $Mdocdate$
.Dt RAXPOL_RAY_HDRS 1
.Os UNIX
.Sh NAME
.Nm raxpol_ray_hdrs
.Nd Print RaXPol ray headers
.Sh SYNOPSIS
.Nm raxpol_ray_hdrs
.Op Fl V
.Op Fl a
.Op Fl l
.Op Fl h Ar angle
.Op Fl s Ar start
.Op Fl c Ar count
.Ar raxpol_file
.Sh DESCRIPTION
.Nm raxpol_ray_hdrs
prints ray headers from RaXPol moment file
.Ar raxpol_file .
.Sh OPTIONS
.Bl -tag -width angle
.It Fl V
Print version information and exit.
.It Fl a
requests abbreviated output.
.It Fl l
indicates
.Ar raxpol_file
uses legacy (2011) format.
.It Fl h Ar angle
sets radar heading to
.Ar angle ,
overriding GPS heading, which might impose spurious variations.
.It Fl s Ar start
specifies index of first ray to print. First ray in file has index 0.
.It Fl c Ar count
specifies number of rays to print.
.El
.Sh OUTPUT FORMAT
Default output starts with the file header, but not the first and last ray
headers, as described in
.Xr raxpol_file_hdr 1 .
Then for each ray, output has form:
.Bd -literal -offset indent
\fBray\fP \fIindex\fP
\fBtimestamp_seconds\fP \fIinteger\fP
\fBtimestamp_useconds\fP \fIinteger\fP
\fBtimestamp_cal\fP \fIYYYY/MM/DD HH:MM:SS.SS\fP
\fBradar_temperatures\fP \fItemp temp temp temp\fP
\fBinclinometer_roll\fP \fInumber\fP
\fBinclinometer_pitch\fP \fInumber\fP
\fBfuel_sensor\fP \fInumber\fP
\fBcpu_temperature\fP \fInumber\fP
\fBpedestal_scan_type\fP \fIcode\fP \fI(name)\fP
\fBtx_power\fP \fInumber\fP
\fBosc_lock\fP \fInumber\fP
\fBaz\fP \fIangle_degrees\fP
\fBel\fP \fIangle_degrees\fP
\fBaz_vel\fP \fInumber\fP
\fBelev_vel\fP \fInumber\fP
\fBaz_current\fP \fInumber\fP
\fBelev_current\fP \fInumber\fP
\fBvolume_count\fP \fIinteger\fP
\fBsweep_count\fP \fIinteger\fP
\fBflags\fP \fInumber\fP
\fBlat_ref\fP \fBN\fPor\fBS\fP
\fBlat\fP \fIlatitude_degrees\fP
\fBlon_ref\fP \fBW\fPor\fBE\fP
\fBlon\fP \fIlongitude_degrees\fP
\fBalt\fP \fIalitude_meters_asl\fP
\fBhdg_ref\fP \fBT\fPor\fBM\fP
\fBhdg\fP \fIdegrees\fP
\fBspeed\fP \fInumber\fP
\fBnmea_msg_gpgga\fP \fIstring\fP
\fButc_time_sec\fP \fIinteger\fP
\fButc_time_usec\fP \fIinteger\fP
\fBdata_type\fP \fIcode\fP
\fBdata_size\fP \fIinteger\fP
.Ed
With
.Fl a
option, each ray has one line of output of form:
.Bd -literal -offset indent
\fBray\fP \fIindex\fP \fIYYYY/MM/DD\fP \fIHH:MM:SS.SS\fP \fBlon\fP \fIdeg\fP \fBlat\fP \fIdeg\fP \fBaz\fP \fIdeg\fP \fBel\fP \fIdeg\fP
.Ed
.Sh ENVIRONMENT
.Ev RAXPOL_OLD_FMT
set and non-zero is equivalent to
.Fl l .
.Sh AUTHOR
.An Gordon Carrie
.Mt dev0@trekix.net


RELEASE_NOTES
Wed Feb 11 16:19:57 CST 2015
raxpol_ray_hdrs also prints file offset to ray

20150316161719
Re-wrote findswps.
Ray header printout includes date as well as time

20150428170551
findswps --
min_span argument can take separate values for PPI and RHI.
Cleaned out dead and deprecated code from findswps.c.
Reformatted output of findswps command for easier searching.
--
raxpol_lib --
Added blocks for 2011 and older raxpol files.
Reformatted output of header print functions for better compatibility with awk.
Added ability to read SPP Sum Power rays.
Various bug fixes. More checking for NULL pointers in functions.
--
raxpol_file_hdr --
prints headers for first and last ray.
Accepts -o option indicating old format files.
--
raxpol_ray_hdr --
Added options.  -a requests abbreviated output.  -h sets heading, overriding
GPS heading in ray headers.  -o input file is "old" (2011) format. Start and
end ray times desired are now specified with -s and -e options, respectively.
--
raxpol_to_sweep --
Accepts separate minimum spans for PPI and RHI.
Is more organinized.
--
tm_calc library has a resolution parameter
raxpol_to_dorade.awk checked in. Makes partial DORADE sweep right now.
raxpol_ray_hdrs utility accepts a time range.

20150507150001
Added RaXPol_Noise function.
Split RaXPol_ZH and RaXPol_ZV into RaXPol_ZH, RaXPol_ZV, RaXPol_DBZH,
and RaXPol_DBZV.
raxpol_file_hdr utility prints offsets to first and last rays.
raxpol_lib --
Created separate raxpol_z and raxpol_dbz functions instead of doing everything
in raxpol_z.
--
raxpol_dat --
utility only calculates and prints requested fields.
raxpol_dat computes moments locally instead of using functions in raxpol_lib,
which are now redundant. RaXPol moments have too many interdependencies for
modular calculations in the current set up.
--

20150925120929
strlcpy.h --
    Disabled global strlcpy before local declaration. This prevents interference
    from Apple's version of strlcpy in their stupid "standard" library.
--
raxpol.h, raxpol_lib.h, raxpol_dat.c --
    Encapsulated the various server modes and recorded moments into RaXPol_Data
    struct.  Consolidated the input and output functions into a smaller set that
    interfaces with the struct. Rewrote raxpol_dat.c with the new interface.
--

20151005153302
findswps.c --
    Added more tests to still_rhi and still_ppi functions.
    Capitalized file scope variables.
--
raxpol.h, raxpol_lib.c, raxpol_dat.c --
    Functions that convert input fields to output are now methods of RaXPol_Data
    struct instead of global funtions.
--

20151005170735
findswps.c --
    Some global variables are now local.
--

20151006200255
raxpol_lib.c --
    Filled in stub functions for single pulse pair and single pulse pair sum
    power output fields.
--

20151013183955
Checking in sweep_img.c, raxpol_swp_idx, raxpol_swp_dat, and raxpol_swp_dat.awk.
Removed raxpol_to_sweep, replace with raxpol_swp_idx.
Checking in geog_proj.c geog_proj.h get_colors.c get_colors.h.
raxpol_lib.c --
    Abbreviated ray header output now includes truck heading.
--
raxpol_dat.c --
    Application no longer prints file header. Printing ray headers is optional.
    Moment names are printed only if there is more than one moment.
--
findswps.c --
    First sweep is index 0, not 1;
--

20151014134133
geog_proj.c --
    GeogProjLonLatToXY and GeogProjXYToLonLat set return arguments to NAN on
    failure.
--
sweep_img.c --
    Added functions to compute gate corners for filled sweeps.
--

20151014190433
findswps, raxpol_swp_idx --
    Now accepts -y option to specify scan type sought, "ppi", "rhi", or "all".
    raxpol_swp_idx uses it.
--

20151102113355
Checking in geog, tm_calc, pisa, color_legend, raxpol_swps_svg, sweep_limits
applications.
Checking in str library utility.
raxpol_ray_hdrs --
    Added ray count option
--
findswps --
    -y option is now -i. This allows the option to pass in
    raxpol_swps_svg, where -y option specifies a style sheet.
--
raxpol_swp_dat, raxpol_swp_dat.awk, raxpol_swp_idx --
    Removed. Not needed.
--
sweep_img --
    Keys end with ":", like pisa.
--
raxpol_lib.c --
    Abbreviated ray header output includes Julian ray time in addition to
    calendar ray time.
--

20151102171827
raxpol_ray_hrs --
    Just takes ray start index and count, or just prints all rays.
    Time search will be a separate application.
--
raxpol_lib.c --
    Shortened abbreviated header output.
--

20151103180433
raxpol_dat --
    Replaced offset argument with index to first ray.
--
sweep_img --
    (bug fix) Corrected order in which gate corners are printed.
    (bug fix) Failure to convert degrees to radians.
--
raxpol_swps_svg --
    Command line options can specify separate bounds for PPI and RHI.
    Adjusted for new abbreviated header output.
--

20151104170343
unix_defs.h --
    c99 is optional.
--
Makefile, raxpol_script_links.svg --
    Path to share directory is substituted into svg script links file.
--

20151104182919
Checking in share directory and contents.
raxpol_swps_svg --
    Color file is now a color directory. Application searches this directory
    for a file with name based on the data type.
    Several variables can now be set either with environment variable or
    command line option.
--

20151105134619
raxpol_dat --
    Always prints moment names, even for only one moment.
--
raxpol_swps_svg --
    Clarified documentation. Strip out moment name from raxpol_dat output.
    sweep_img does not need it.
--

20151105195015
raxpol_dat --
    Now requires starting ray index and ray count. Output consists of moment name
    followed by data for ray count rays, then next moment name followed by data
    for ray count rays, and so on.
--
raxpol_swps_svg --
    Reads the revised raxpol_dat output. Makes sweeps for several moments in
    one block. Desired performance boost failed to show.
--

20151105195715
raxpol_dat, raxpol_swps_svg --
    Restoring from 20151105134619 release. It is simpler and performs as well

src/raxpol_ray_hdrs.c
/*
   -	raxpol_ray_hdrs.c --
   -		This program defines an application prints file header
   -		and ray headers from a raxpol file.
   .
   .	Usage:
   .		raxpol_ray_hdrs [-a] [-l] [-h angle] [-s start] [-c count]
   .			[raxpol_file]
   .
   .	Options:
   .		-a requests abbreviated output.
   .		-l input file is "old" (2011) format.
   .		-h sets heading, overriding GPS heading in ray headers.
   .		-s index of first ray to read. First ray is 0.
   .		-c number of rays to read
   .
   .	Non-zero RAXPOL_OLD_FMT environment variable is same as -l.
   .
   .	Reading proceeds until count or end of file.
   .
   .	Ray headers are written to standard output.
   .
   .	Copyright (c) 2012, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
   .
   .	$Revision: 1.8 $ $Date: 2015/02/10 21:07:56 $
 */

#include "unix_defs.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <limits.h>
#include <errno.h>
#include "raxpol.h"

#define RAXPOL_OLD_FMT "RAXPOL_OLD_FMT"

int main(int argc, char *argv[])
{
    char *argv0;			/* Name of the executable, for error
					   messages. */
    int c;				/* Index into argv */
    extern char *optarg;		/* See getopt (3) */
    double dflt_hdg;			/* Default heading */
    char *raxpol_fl_nm;			/* RaXPol file path */
    FILE *raxpol_fl;			/* RaXPol file */
    long r0, num_rays;			/* First ray, number of rays to read */
    long num_rays_max;			/* Number of rays from r0 to EOF */
    off_t o0;				/* Offset to start of first ray */
    off_t o;				/* Offset somewhere in raxpol_fl */
    off_t ray_sz;			/* Size in file of one ray */
    int abbrv;				/* If true, abbreviate */
    struct RaXPol_File_Hdr file_hdr;	/* Header from the RaXPol file */
    struct RaXPol_Ray_Hdr ray_hdr;
    long r;

    argv0 = argv[0];
    abbrv = 0;
    r0 = 0;
    num_rays = LONG_MAX;
    if ( atoi(getenv(RAXPOL_OLD_FMT) ? getenv(RAXPOL_OLD_FMT) : "0") ) {
	RaXPol_Old_Fmt();
    }
    while ((c = getopt(argc, argv, ":Valh:s:c:")) != -1) {
	switch(c) {
	    case 'V':
		printf("%s\n", RAXPOL_MOMENT_VERSION);
		exit(EXIT_SUCCESS);
		break;
	    case 'a':
		abbrv = 1;
		break;
	    case 'l':
		RaXPol_Old_Fmt();
		break;
	    case 'h':
		if ( sscanf(optarg, "%lf", &dflt_hdg) != 1 ) {
		    fprintf(stderr, "%s: expected float value for default "
			    "heading, got %s\n", argv0, optarg);
		    exit(EXIT_FAILURE);
		}
		RaXPol_Set_Hdg(dflt_hdg);
		break;
	    case 's':
		if ( sscanf(optarg, "%ld", &r0) != 1 ) {
		    fprintf(stderr, "%s: expected integer for index of first "
			    "ray, got %s\n", argv0, optarg);
		    exit(EXIT_FAILURE);
		}
	    case 'c':
		if ( sscanf(optarg, "%ld", &num_rays) != 1 ) {
		    fprintf(stderr, "%s: expected integer for desired ray "
			    "count, got %s\n", argv0, optarg);
		    exit(EXIT_FAILURE);
		}
		break;
	    case '?':
		fprintf(stderr, "Usage: %s [-a] [-s YYYYMMDD-HHMMSS] "
			"[-e YYYYMMDD-HHMMSS] [raxpol_file]\n", argv0);
		exit(EXIT_FAILURE);
		break;
	}
    }
    if ( optind == argc ) {
	raxpol_fl_nm = "-";
    } else {
	raxpol_fl_nm = argv[optind];
    }
    if ( strcmp(raxpol_fl_nm, "-") == 0 ) {
	raxpol_fl = stdin;
    } else if ( !(raxpol_fl = fopen(raxpol_fl_nm, "r")) ) {
	fprintf(stderr, "%s: could not open %s for reading.\n",
		argv0, raxpol_fl_nm);
	exit(EXIT_FAILURE);
    }
    RaXPol_Init_File_Hdr(&file_hdr);
    if ( !RaXPol_Read_File_Hdr(&file_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read file header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    if ( (o0 = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine position in file.\n%s\n",
		argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }

    /* Determine ray size and num_rays_max */
    RaXPol_Init_Ray_Hdr(&ray_hdr);
    if ( !RaXPol_Read_Ray_Hdr(&ray_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read header for first ray.\n",
		argv0);
	exit(EXIT_FAILURE);
    }
    if ( (o = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine position in file.\n%s\n",
		argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }
    ray_sz = o - o0 + ray_hdr.data_size;
    if ( fseeko(raxpol_fl, 0, SEEK_END) == -1
	    || (o = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not position at end of file.\n"
		"%s\n", argv0, strerror(errno));
	exit(EXIT_FAILURE);
    }
    num_rays_max = (o - o0) / ray_sz - r0;
    if ( num_rays > num_rays_max ) {
	num_rays = num_rays_max;
    }

    /* Move to first ray */
    if ( r0 > 0 ) {
	if ( fseeko(raxpol_fl, o0 + r0 * ray_sz, SEEK_SET) == -1 ) {
	    fprintf(stderr, "%s: could not position at start of first ray.\n"
		    "%s\n", argv0, strerror(errno));
	    exit(EXIT_FAILURE);
	}
    } else {
	if ( fseeko(raxpol_fl, o0, SEEK_SET) == -1) {
	    fprintf(stderr, "%s: could not position at end of file.\n"
		    "%s\n", argv0, strerror(errno));
	    exit(EXIT_FAILURE);
	}
    }

    /* Print ray headers */
    if ( !abbrv ) {
	printf("Reading %s\n",
		(raxpol_fl == stdin) ? "standard input" : raxpol_fl_nm);
	printf("File header:\n");
	RaXPol_FPrintf_File_Hdr(&file_hdr, stdout);
    }
    for (r = r0;
	    r < r0 + num_rays && RaXPol_Read_Ray_Hdr(&ray_hdr, raxpol_fl);
	    r++) {
	if ( abbrv ) {
	    printf("ray %-9ld ", r);
	    RaXPol_FPrint_Abbrv_Ray_Hdr(&ray_hdr, stdout);
	} else {
	    printf("******************* ray *******************\n");
	    printf("ray %ld\n", r);
	    RaXPol_FPrint_Ray_Hdr(&ray_hdr, stdout);
	}
	if ( fseeko(raxpol_fl, ray_hdr.data_size, SEEK_CUR) == -1 ) {
	    fprintf(stderr, "%s: could not skip ray data\n%s\n",
		    argv0, strerror(errno));
	    exit(EXIT_FAILURE);
	}
    }
    if ( feof(raxpol_fl) || ferror(raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read ray header for ray %ld of %s\n",
		argv0, r, raxpol_fl_nm);
	exit(EXIT_FAILURE);
    }

    return EXIT_SUCCESS;
}

